signac/contrib/project.py

```diff
@@ -176,7 +173,8 @@ def __init__(self, config=None, _ignore_schema_version=False):
 
     def __str__(self):
         """Return the project's id."""
-        return str(self.id)
+        # TODO: Is this the string representation that we want?
+        return os.path.split(self._root_directory)[-1]
```

Comment:
I think maybe `Project(self.root_directory())` would be good. The `__repr__` for the class uses `get_project`, which I find a bit strange. While we don't tell users to instantiate projects directly, I'm not sure that it makes sense to look like that in the repr. I'd like the repr and the str to match up relatively closely here. Thoughts?

Reply:
The invariant I sometimes advocate for is `eval(repr(obj)) == obj`. We use this in freud, for instance, to ensure that constructors align with the object representation. I recognize it is more awkward for signac because projects are constructed via factory methods. I think the repr is important and would like other opinions on this.

Reply:
Exactly, the use of factories make that invariant seem a little awkward here.

Reply:
I explicitly implemented the `__repr__()` to use the factory function to ensure the aforementioned invariant. I would advocate to change the constructor signature with all the other breaking changes for 2.0 and construct a `Project` directly from the root directory. That should be sufficient, considering that it is **the** unique identifier. User should still use `get_project()` for project discovery.

With that in mind I would suggest the following:
```python
>>>> str(project)
'<signac.Project at "." with 42 jobs>'
>>> repr(project)
'Project("/absolute/path/to/project")'
```
What do you think?

Reply:
Yes I like that idea. @bdice any objections?

Reply:
One related question here that we should think about is whether the project constructor would have the semantics of `get_project` or `init_project`. Will the construct create a project if one does not already exist in a directory?

Reply:
Strongly against including discovery with the constructor, that just leads to messy code and also breaks the invariant from above. Is `Project('/path/to/project/sub-dir', discover=True) == Project('/path/to/project')`?

I'd also caution against side-effects on object construction. We had this for jobs initially and thankfully moved away from that. I see two viable approaches:
1. We require the Project to "exist" before we can obtain the corresponding object for it. So, `Project(/path/to/project')` would raise an exception in case the `/paht/to/project/.signac` does not exist.
2. We don't require the project to exist, and create it on the first write-operation. That's similar to how jobs work: `Project('/path/to/project').init()`.
I lean toward approach 2.

Reply:
I agree with no discovery. At first glance I like approach 2 as well, although I'd have to think about the other implications. Since our plan is to tighten the job/project linkage, we'd probably also want things like job creation to imply project creation and so on. There would likely be additional such implied behaviors.